src/backoff.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+// https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+
+use rand::{thread_rng, Rng};
+
+pub trait Backoff: Clone + Copy + Send + Sync + 'static {
+    fn next_delay_ms(&mut self) -> Option<u64>;
+
+    fn exponential_delay_ms(base_delay_ms: u64, max_delay_ms: u64, attempts: u32) -> u64 {
+        max_delay_ms.min(2u64.pow(attempts) * base_delay_ms)
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct BackoffState {
```

Comment:
I think it might be a better design to put only mutable things in the state. Immutable fields can be saved in the backoff type directly.